insights/parsers/podman_list.py

```diff
@@ -0,0 +1,127 @@
+stored.
+
+Each row is stored as a dictionary, keyed on the header fields. The data is
+available in two formats:
+
+* The old format is a list of row dictionaries.
+* The new format stores each dictionary in a dictionary keyed on the value of
+  a given field, given by the subclass.
+
+Sample output of command ``/usr/bin/podman images --all --no-trunc --digests``::
+
+    REPOSITORY                           TAG                 DIGEST              IMAGE ID                                                           CREATED             SIZE
+    rhel7_imagemagick                    latest              <none>              882ab98aae5394aebe91fe6d8a4297fa0387c3cfd421b2d892bddf218ac373b2   4 days ago          785.4 MB
+    rhel6_nss-softokn                    latest              <none>              dd87dad2c7841a19263ae2dc96d32c501ee84a92f56aed75bb67f57efe4e48b5   5 days ago          449.7 MB
+
+
+Sample output of command ``/usr/bin/podman ps --all --no-trunc --size``::
+
+    CONTAINER ID                                                       IMAGE                                                              COMMAND                                            CREATED             STATUS                        PORTS                  NAMES               SIZE
+    95516ea08b565e37e2a4bca3333af40a240c368131b77276da8dec629b7fe102   bd8638c869ea40a9269d87e9af6741574562af9ee013e03ac2745fb5f59e2478   "/bin/sh -c 'yum install -y vsftpd-2.2.2-6.el6'"   51 minutes ago      Exited (137) 50 minutes ago                          tender_rosalind     4.751 MB (virtual 200.4 MB)
+    03e2861336a76e29155836113ff6560cb70780c32f95062642993b2b3d0fc216   rhel7_httpd                                                        "/usr/sbin/httpd -DFOREGROUND"                     45 seconds ago      Up 37 seconds                 0.0.0.0:8080->80/tcp   angry_saha          796 B (virtual 669.2 MB)
+
+Examples:
+    >>> images.rows[0]['REPOSITORY']
+    'rhel6_vsftpd'
+    >>> images.rows[1]['SIZE']
+    '785.4 MB'
+    >>> images.data['rhel6_vsftpd']['CREATED']
+    '37 minutes ago'
+    >>> containers.rows[0]['STATUS']
+    'Up 37 seconds'
+    >>> containers.data['tender_rosalind']['STATUS']
+    'Exited (137) 18 hours ago'
+"""
+import re
+from insights import CommandParser, parser
+from insights.parsers import SkipException
+from insights.specs import Specs
+
+
+class PodmanList(CommandParser):
+    """
+    A general class for parsing tabular podman list information.  Parsing
+    rules are:
+
+    * The first line is the header line.
+    * The other lines are data lines.
+    * All fields line up vertically.
+    * Fields are separated from each other by at least three spaces.
+    * Some fields can contain nothing, and this is shown as spaces, so we
+      need to catch that and turn it into None.
+
+    Why not just use hard-coded fields and columns?  So that we can adapt to
+    different output lists.
+    """
+    key_field = None
+
+    def parse_content(self, content):
```

Comment:
A utility function was introduced to the core for this kind of content `parse_fixed_table` and better re-name the `self.data` to a name which could represent the content.

E.g:

```python
from insights.parsers import SkipException, parse_fixed_table

class PodmanList(CommandParser):
    key_field = ''
    attr_name = ''

    def parse_content(self, content):
        self.rows = parse_fixed_table(content,
                heading_ignore=[self.key_field])

        if not self.rows:
            raise SkipException

        data = {}
        for row in self.rows:
            k = row.get(self.key_field)
            if k is not None and k != '<none>':
                data[k] = row
        setattr(self, self.attr_name, data)

@parser(Specs.podman_list_images)
class PodmanListImages(PodmanList):
    """
    Handle the list of podman images using the PodmanList parser class.
    """
    key_field = 'REPOSITORY'
    attr_name = 'images'


@parser(Specs.podman_list_containers)
class PodmanListContainers(PodmanList):
    """
    Handle the list of podman containers using the PodmanList parser class.
    """
    key_field = 'NAMES'
    attr_name = 'containers'
```

Please have a try.

Reply:
And please add corresponding docstrings for the attributes and exceptions by referring to https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html

See also:
https://github.com/RedHatInsights/insights-core/blob/fc703496ccbe97944a69f196f0f750dbc030fa74/insights/parsers/grubby.py#L34

https://github.com/RedHatInsights/insights-core/blob/fc703496ccbe97944a69f196f0f750dbc030fa74/insights/parsers/grubby.py#L38

Reply:
> A utility function was introduced to the core for this kind of content `parse_fixed_table` and better re-name the `self.data` to a name which could represent the content.
> 
> E.g:
> 
> ```python
> from insights.parsers import SkipException, parse_fixed_table
> 
> class PodmanList(CommandParser):
>     key_field = ''
>     attr_name = ''
> 
>     def parse_content(self, content):
>         self.rows = parse_fixed_table(content,
>                 heading_ignore=[self.key_field])
> 
>         if not self.rows:
>             raise SkipException
> 
>         data = {}
>         for row in self.rows:
>             k = row.get(self.key_field)
>             if k is not None and k != '<none>':
>                 data[k] = row
>         setattr(self, self.attr_name, data)
> 
> @parser(Specs.podman_list_images)
> class PodmanListImages(PodmanList):
>     """
>     Handle the list of podman images using the PodmanList parser class.
>     """
>     key_field = 'REPOSITORY'
>     attr_name = 'images'
> 
> 
> @parser(Specs.podman_list_containers)
> class PodmanListContainers(PodmanList):
>     """
>     Handle the list of podman containers using the PodmanList parser class.
>     """
>     key_field = 'NAMES'
>     attr_name = 'containers'
> ```
> 
> Please have a try.

In this case, it won't work since it tries to split "IMAGE ID" column as "IMAGE" and "ID" columns.

> (Pdb) result.rows[0]
{'REPOSITORY': 'rhel6_vsftpd', 'TAG': 'latest', 'DIGEST': '<none>', 'IMAGE': '412b68', 'ID': '4338a1178f0e5ad68a5fd00df01a10a18495959398b2cf92c2033d3d02', 'CREATED': '37 minutes ago', 'SIZE': '459.5 MB'}

Reply:
> And please add corresponding docstrings for the attributes and exceptions by referring to https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html
> 
> See also:
> https://github.com/RedHatInsights/insights-core/blob/fc703496ccbe97944a69f196f0f750dbc030fa74/insights/parsers/grubby.py#L34
> 
> https://github.com/RedHatInsights/insights-core/blob/fc703496ccbe97944a69f196f0f750dbc030fa74/insights/parsers/grubby.py#L38

Done

Reply:
It works on for podman list images and not for podman list containers.

Reason:
The implementation of parse_fixed_table is not up to the mark or doesn't cover all the scenarios appropriately.

**Reason**:
1. What if someone wants to search for strings which are not at the begging of line. It has a hardcode binding that "

> heading_ignore (list): Optional list of strings to search for at beginning of line

2. What if someone chooses a string containing white space character inside it as `heading_ignore` and same has been added to `header_substitute`?

>     key_field = 'CONTAINER ID'
>     attr_name = 'containers'
>     header_substitute = [("CONTAINER ID", "CONTAINER_ID")]

It doesn't work as expected since it tries to find the line containing `key_field` and then immediately normalizes the values passed in `header_substitute` resulting `k` variable having value as `None`

> $ /opt/insights/insights-core/insights/parsers/podman_list.py(88)parse_content()
> Current line: k = row.get(self.key_field)
> (Pdb) self.key_field
> 'CONTAINER ID'
> (Pdb) row
> {'CONTAINER_ID': '74a2ee87b0c7', 'IMAGE': '192.168.24.1:8787/rhosp15/openstack-ceilometer-central:20190819.1', 'COMMAND': 'dumb-init --singl...', 'CREATED': '2 weeks ago', 'STATUS': 'Exited (0) 2 weeks ago', 'PORTS': '', 'NAMES': 'ceilometer_gnocchi_upgrade'}

I could put a whitelisting to hack this however it won't be a good way to implement in such since I am trying to get a generalized parsers for podman list images and podman list containers similar to what has been done for docker

Reply:
The original implementation of the "docker list" is implemented before introducing the `parse_fixed_table`.  The reason for suggesting you changing to the `parse_fixed_table` is that this interface removes the duplicated code and provides an easy to maintain interface. By modifying the sample code I gave above, it would satisfy the requirements/concerns you mentioned above:

~~~diff
diff --git a/insights/parsers/podman_list.py b/insights/parsers/podman_list.py
index a2b30934..410cbc5c 100644
--- a/insights/parsers/podman_list.py
+++ b/insights/parsers/podman_list.py
@@ -20,35 +20,11 @@ available in two formats:
 * The new format stores each dictionary in a dictionary keyed on the value of
   a given field, given by the subclass.

-Sample output of command ``/usr/bin/podman images --all --no-trunc --digests``::
-
-    REPOSITORY                           TAG                 DIGEST              IMAGE ID                                                           CREATED             SIZE
-    rhel7_imagemagick                    latest              <none>              882ab98aae5394aebe91fe6d8a4297fa0387c3cfd421b2d892bddf218ac373b2   4 days ago          785.4 MB
-    rhel6_nss-softokn                    latest              <none>              dd87dad2c7841a19263ae2dc96d32c501ee84a92f56aed75bb67f57efe4e48b5   5 days ago          449.7 MB
-
-
-Sample output of command ``/usr/bin/podman ps --all --no-trunc --size``::
-
-    CONTAINER ID                                                       IMAGE                                                              COMMAND                                            CREATED             STATUS                        PORTS                  NAMES               SIZE
-    95516ea08b565e37e2a4bca3333af40a240c368131b77276da8dec629b7fe102   bd8638c869ea40a9269d87e9af6741574562af9ee013e03ac2745fb5f59e2478   "/bin/sh -c 'yum install -y vsftpd-2.2.2-6.el6'"   51 minutes ago      Exited (137) 50 minutes ago                          tender_rosalind     4.751 MB (virtual 200.4 MB)
-    03e2861336a76e29155836113ff6560cb70780c32f95062642993b2b3d0fc216   rhel7_httpd                                                        "/usr/sbin/httpd -DFOREGROUND"                     45 seconds ago      Up 37 seconds                 0.0.0.0:8080->80/tcp   angry_saha          796 B (virtual 669.2 MB)
-
-Examples:
-    >>> images.rows[0]['REPOSITORY']
-    'rhel6_vsftpd'
-    >>> images.rows[1]['SIZE']
-    '785.4 MB'
-    >>> images.data['rhel6_vsftpd']['CREATED']
-    '37 minutes ago'
-    >>> containers.rows[0]['STATUS']
-    'Up 37 seconds'
-    >>> containers.data['tender_rosalind']['STATUS']
-    'Exited (137) 18 hours ago'
+
 """
-import re
 from insights import CommandParser, parser
-from insights.parsers import SkipException
 from insights.specs import Specs
+from insights.parsers import SkipException, parse_fixed_table


 class PodmanList(CommandParser):
@@ -66,62 +42,78 @@ class PodmanList(CommandParser):
     Why not just use hard-coded fields and columns?  So that we can adapt to
     different output lists.
     """
-    key_field = None
+    key_field = ''
+    heading_ignore = []
+    attr_name = ''
+    substitutions = []

     def parse_content(self, content):
-        """
-        Parse the lines given into a list of dictionaries for each row.  This
-        is stored in the ``rows`` attribute.
-
-        If the ``key_field`` property is set, use this to key a ``data``
-        dictionary attribute.
-        """
-        self.rows = []
-        if len(content) < 2:
-            raise SkipException("No data.")
-
-        # Parse header, remembering column numbers for data capture.  We use
-        # a finditer to get the positions, and we find by field rather than
-        # splitting on three or more spaces because of this.
-        headers = []
-        field_re = re.compile(r'\w+(\s\w+)*')
-        for match in field_re.finditer(content[0]):
-            headers.append({'name': match.group(), 'start': match.start()})
-
-        # Parse the rest of the line.  Each field starts at the column
-        # given by the header and ends with at least three spaces.
-        for line in content[1:]:
-            # I think the dictionary comprehension version of this is too
-            # complicated for words :-)
-            row = {}
-            for header in headers:
-                value = line[header['start']:].split('   ', 1)[0]
-                if value == '':
-                    value = None
-                row[header['name']] = value
-            self.rows.append(row)
-
-        # If we have a key_field set, construct a data dictionary on it.
-        # Note that duplicates will be overwritten, but we ignore '<none>'.
-        if self.key_field and self.key_field in self.rows[0]:
-            self.data = {}
-            for row in self.rows:
-                k = row[self.key_field]
-                if k is not None and k != '<none>':
-                    self.data[k] = row
+        if not (self.key_field and self.attr_name):
+            raise NotImplementedError("'key_field' or 'attr_name' is not defined")
+
+        self.rows = parse_fixed_table(content,
+                heading_ignore=self.heading_ignore,
+                header_substitute=self.substitutions)
+
+        if not self.rows:
+            raise SkipException('No data.')
+
+        data = {}
+        for row in self.rows:
+            k = row.get(self.key_field)
+            for sub in self.substitutions:
+                row[sub[0]] = row.pop(sub[1])
+            if k is not None and k != '<none>':
+                data[k] = row
+        setattr(self, self.attr_name, data)


 @parser(Specs.podman_list_images)
 class PodmanListImages(PodmanList):
     """
     Handle the list of podman images using the PodmanList parser class.
+
+    Sample output of command ``podman images --all --no-trunc --digests``::
+
+        REPOSITORY                           TAG                 DIGEST              IMAGE ID                                                           CREATED             SIZE
+        rhel6_vsftpd                         latest              <none>              412b684338a1178f0e5ad68a5fd00df01a10a18495959398b2cf92c2033d3d02   37 minutes ago      459.5 MB
+        rhel7_imagemagick                    latest              <none>              882ab98aae5394aebe91fe6d8a4297fa0387c3cfd421b2d892bddf218ac373b2   4 days ago          785.4 MB
+        rhel6_nss-softokn                    latest              <none>              dd87dad2c7841a19263ae2dc96d32c501ee84a92f56aed75bb67f57efe4e48b5   5 days ago          449.7 MB
+
+    Examples:
+        >>> images.rows[0]['REPOSITORY']
+        'rhel6_vsftpd'
+        >>> images.rows[1]['SIZE']
+        '785.4 MB'
+        >>> images.images['rhel6_vsftpd']['CREATED']
+        '37 minutes ago'
     """
     key_field = 'REPOSITORY'
+    heading_ignore = [key_field]
+    attr_name = 'images'
+    substitutions = [("IMAGE ID", "IMAGE_ID")]


 @parser(Specs.podman_list_containers)
 class PodmanListContainers(PodmanList):
     """
     Handle the list of podman containers using the PodmanList parser class.
+
+    Sample output of command ``podman ps --all --no-trunc --size``::
+
+        CONTAINER ID                                                       IMAGE                                                              COMMAND                                            CREATED             STATUS                        PORTS                  NAMES               SIZE
+        03e2861336a76e29155836113ff6560cb70780c32f95062642993b2b3d0fc216   rhel7_httpd                                                        "/usr/sbin/httpd -DFOREGROUND"                     45 seconds ago      Up 37 seconds                 0.0.0.0:8080->80/tcp   angry_saha          796 B (virtual 669.2 MB)
+        95516ea08b565e37e2a4bca3333af40a240c368131b77276da8dec629b7fe102   bd8638c869ea40a9269d87e9af6741574562af9ee013e03ac2745fb5f59e2478   "/bin/sh -c 'yum install -y vsftpd-2.2.2-6.el6'"   51 minutes ago      Exited (137) 50 minutes ago                          tender_rosalind     4.751 MB (virtual 200.4 MB)
+
+    Examples:
+        >>> containers.rows[0]['NAMES']
+        'angry_saha'
+        >>> containers.rows[0]['STATUS']
+        'Up 37 seconds'
+        >>> containers.containers['tender_rosalind']['STATUS']
+        'Exited (137) 18 hours ago'
     """
     key_field = 'NAMES'
+    heading_ignore = ['CONTAINER']
+    attr_name = 'containers'
+    substitutions = [("CONTAINER ID", "CONTAINER_ID")]
diff --git a/insights/parsers/tests/test_podman_list.py b/insights/parsers/tests/test_podman_list.py
index 49289dc6..b93d8cec 100644
--- a/insights/parsers/tests/test_podman_list.py
+++ b/insights/parsers/tests/test_podman_list.py
@@ -43,11 +43,11 @@ def test_podman_list_images():
     assert result.rows[2].get("TAG") == "latest"
     assert result.rows[2].get("IMAGE ID") == 'dd87dad2c7841a19263ae2dc96d32c501ee84a92f56aed75bb67f57efe4e48b5'

-    assert result.data['rhel6_vsftpd']['CREATED'] == '37 minutes ago'
+    assert result.images['rhel6_vsftpd']['CREATED'] == '37 minutes ago'
     # Same data in both accessors
-    assert result.data['rhel6_vsftpd'] == result.rows[0]
+    assert result.images['rhel6_vsftpd'] == result.rows[0]
     # Can't list repositories if they don't have a repository name
-    assert '<none>' not in result.data
+    assert '<none>' not in result.images


 def test_podman_list_containers():
@@ -61,21 +61,22 @@ def test_podman_list_containers():
     assert result.rows[1].get("CONTAINER ID") == "95516ea08b565e37e2a4bca3333af40a240c368131b77276da8dec629b7fe102"
     assert result.rows[1].get("COMMAND") == '"/bin/sh -c \'yum install -y vsftpd-2.2.2-6.el6\'"'
     assert result.rows[1]['STATUS'] == 'Exited (137) 18 hours ago'
-    assert result.rows[1].get("PORTS") is None
+    assert result.rows[1].get("PORTS") is ''

-    assert sorted(result.data.keys()) == sorted(['angry_saha', 'tender_rosalind'])
-    assert result.data['angry_saha'] == result.rows[0]
-    assert result.data['tender_rosalind'] == result.rows[1]
+    assert sorted(result.containers.keys()) == sorted(['angry_saha', 'tender_rosalind'])
+    assert result.containers['angry_saha'] == result.rows[0]
+    assert result.containers['tender_rosalind'] == result.rows[1]


 def test_podman_list_images_no_data():
     with pytest.raises(SkipException) as ex:
-        podman_list.PodmanList(context_wrap(PODMAN_LIST_IMAGES_NO_DATA))
+        podman_list.PodmanListImages(context_wrap(PODMAN_LIST_IMAGES_NO_DATA))
     assert 'No data.' in str(ex)


 def test_podman_undefined_key_field():
-    assert podman_list.PodmanList(context_wrap(PODMAN_LIST_CONTAINERS)).key_field is None
+    with pytest.raises(NotImplementedError):
+        assert podman_list.PodmanList(context_wrap(PODMAN_LIST_CONTAINERS))

~~~